include/merlin/core_kernels.cuh

```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
+*/
+template <class K, class V, class M>
+__global__ void read_or_write_kernel(V** __restrict table_value_addrs, V* __restrict param_values,
+                            const bool* mask, const int* __restrict param_key_index,
+                            const size_t dim, const size_t N) {
+  size_t tid = (blockIdx.x * blockDim.x) + threadIdx.x;
+
+  for (size_t t = tid; t < N; t += blockDim.x * gridDim.x) {
+    int vec_index = int(t / dim);
+    int dim_index = t % dim;
+    int real_key_index =
+        param_key_index != nullptr ? param_key_index[vec_index] : vec_index;
+
+    /// if found, read the value form table, otherwise write it
+    if (table_value_addrs[vec_index] != nullptr) {
+      /// find
+      if (mask[real_key_index]) {
+        param_values[real_key_index * dim + dim_index] = table_value_addrs[vec_index][dim_index];
+      }
+      /// insert
+      else {
+        table_value_addrs[vec_index][dim_index] = param_values[real_key_index * dim + dim_index];
+      }
+    }
+  }
+}
+
+/* 
+ * update with IO operation. This kernel is
+ * usually used for the pure HBM mode for better performance.
+ */
+template <class K, class V, class M, uint32_t TILE_SIZE = 4>
+__forceinline__ __device__ void update_kernel_with_io_core(
+    const Table<K, V, M>* __restrict table, const K* __restrict keys,
+    const V* __restrict values, const M* __restrict metas, const size_t N) {
+  Bucket<K, V, M>* buckets = table->buckets;
+  const size_t bucket_max_size = table->bucket_max_size;
+  const size_t buckets_num = table->buckets_num;
+  const size_t dim = table->dim;
+  int* buckets_size = table->buckets_size;
+
+  auto g = cg::tiled_partition<TILE_SIZE>(cg::this_thread_block());
+  int rank = g.thread_rank();
+
+  for (size_t t = (blockIdx.x * blockDim.x) + threadIdx.x; t < N;
+       t += blockDim.x * gridDim.x) {
+    int key_idx = t / TILE_SIZE;
+    int local_size = 0;
+
+    const K update_key = keys[key_idx];
+    const V* update_value = values + key_idx * dim;
+
+    int key_pos = -1;
+    size_t bkt_idx = 0;
+    size_t start_idx = 0;
+    uint32_t tile_offset = 0;
+
+    Bucket<K, V, M>* bucket = get_key_position<K>(
+        buckets, update_key, bkt_idx, start_idx, buckets_num, bucket_max_size);
+    
+    local_size = buckets_size[bkt_idx];
+
+    auto const found_vote = find_in_bucket<K, V, M, TILE_SIZE>(
+        g, bucket->keys, update_key, tile_offset, start_idx, bucket_max_size);
+
+    if (found_vote) {
+      const int dst_lane = __ffs(found_vote) - 1;
+      key_pos = (start_idx + tile_offset + dst_lane) & (bucket_max_size - 1);
+      V* dst = bucket->vectors + key_pos * dim;
+
+      if (rank == dst_lane) {
+        if (metas != nullptr) {
+          bucket->metas[key_pos].store(metas[key_idx], cuda::std::memory_order_relaxed);
+        }
+        else {
+          M cur_meta = bucket->cur_meta.fetch_add(1) + 1;
+          bucket->metas[key_pos].store(cur_meta, cuda::std::memory_order_relaxed);
+        }
```

Comment:
Why not use `update_meta` kernel directly?

Reply:
OK, I will use `update_meta`. Forgive me forgotten it. :D